src/Stack/PID1.hs

```diff
@@ -0,0 +1,131 @@
+module Stack.PID1
+    ( run
+    ) where
+
+-- This is a valid PID 1 process in Haskell, intended as a Docker
+-- entrypoint. It will handle reaping orphans and handling TERM and
+-- INT signals. This is a work in progress, please do not use it in
+-- production!
+
+import           Control.Concurrent       (forkIO, newEmptyMVar, takeMVar,
+                                           threadDelay, tryPutMVar)
+import           Control.Exception        (assert, catch, throwIO)
+import           Control.Monad            (forever, void)
+import           System.Exit              (ExitCode (ExitFailure), exitWith)
+import           System.IO.Error          (isDoesNotExistError)
+import           System.Posix.Process     (ProcessStatus (..), exitImmediately,
+                                           getAnyProcessStatus)
+import           System.Posix.Signals     (Handler (Catch), Signal,
+                                           installHandler, sigINT, sigKILL,
+                                           sigTERM, signalProcess)
+import           System.Posix.Types       (CPid)
+import           System.Process           (createProcess, env, proc)
+import           System.Process.Internals (ProcessHandle__ (..),
+                                           modifyProcessHandle)
+
+run :: FilePath -- ^ command
+    -> [String] -- ^ arguments
+    -> Maybe [(String, String)] -- ^ env override
+    -> IO a
+run cmd args environment = do
+    (Nothing, Nothing, Nothing, ph) <- createProcess (proc cmd args)
+        { env = environment
+        }
+
+    -- Determine the child PID. We want to exit once this child
+    -- process is dead.
+    p_ <- modifyProcessHandle ph $ \p_ -> return (p_, p_)
+    child <-
+        case p_ of
+            ClosedHandle e -> assert False (exitWith e)
```

Comment:
Is the `assert False` necessary?  exitWith already essentially returns `IO void`